accepted/2024/support-nuget-authentication-plugins-dotnet-tools.md

```diff
@@ -0,0 +1,504 @@
+1. Execute the command `dotnet tool install -g Microsoft.CredentialProvider`.
+1. Run `dotnet restore --interactive` with a private endpoint.
+    It should 'just work', meaning the credential providers installed in step 2 are used during credential acquisition and are used to authenticate against private endpoints.
+
+Upon installation, these global .NET tools are added to the PATH.
+This allows NuGet to easily determine which file in the package should be run at runtime.
+
+### Security considerations
+
+The [.NET SDK docs](https://learn.microsoft.com/dotnet/core/tools/global-tools#check-the-author-and-statistics) clearly state, `.NET tools run in full trust. Don't install a .NET tool unless you trust the author`.
+This is an important consideration for plugin customers when installing NuGet plugins via .NET Tools in the future.
+
+### Technical explanation
+
+#### Plugin discovery
+
+Currently, plugins are discovered through a convention-based directory structure, such as the `%userprofile%/.nuget/plugins` folder on `Windows`.
+For CI/CD scenarios, and for power users, environment variables can be used to override this behavior.
+Note that only absolute paths are allowed when using these environment variables.
+
+- `NUGET_NETFX_PLUGIN_PATHS`: Defines the plugins used by the .NET Framework-based tooling (NuGet.exe/MSBuild.exe/Visual Studio). This takes precedence over `NUGET_PLUGIN_PATHS`.
+- `NUGET_NETCORE_PLUGIN_PATHS`: Defines the plugins used by the .NET Core-based tooling (dotnet.exe).
+This takes precedence over `NUGET_PLUGIN_PATHS`.
+- `NUGET_PLUGIN_PATHS`: Defines the plugins used for the NuGet process, with priority preserved.
+If this environment variable is set, it overrides the convention-based discovery.
+It is ignored if either of the framework-specific variables is specified.
+
+I propose the addition of a `NUGET_DOTNET_TOOLS_PLUGIN_PATHS` environment variable.
+This variable will define the plugins, installed as .NET tools, to be used by both .NET Framework and .NET Core tooling.
+It will take precedence over `NUGET_PLUGIN_PATHS`.
+
+The plugins specified in the `NUGET_DOTNET_TOOLS_PLUGIN_PATHS` environment variable will be used regardless of whether the `NUGET_NETFX_PLUGIN_PATHS` or `NUGET_NETCORE_PLUGIN_PATHS` environment variables are set.
+The primary reason for this is that plugins installed as .NET tools can be executed in both .NET Framework and .NET Core tooling.
+If customers prefer to install NuGet plugins as a [tool-path global tool](https://learn.microsoft.com/dotnet/core/tools/global-tools-how-to-use#use-the-tool-as-a-global-tool-installed-in-a-custom-location), they can set the `NUGET_DOTNET_TOOLS_PLUGIN_PATHS` environment variable.
+This variable should point to the location of the .NET Tool executable that the NuGet Client tooling can invoke when needed.
+
+NuGet should search for files whose names begin with `nuget-plugin-*` by scanning all the directories in the `PATH` environment variable.
```

Comment:
How do we decide in which order the plugins will be used?
I don't foresee multiple plugins to be a common scenario, but it's worth specifying the order.

Reply:
In Visual Studio, the credential providers are invoked in the [following order](https://github.com/NuGet/NuGet.Client/blob/b869c3cc0e0012e297e9f40742daab4168c25d27/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/DefaultVSCredentialServiceProvider.cs#L41-L99):

1. Credential providers installed into [Visual Studio as extensions](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/VsCredentialProviderImporter.cs) (specifically, those that implement IVSCredentialProvider).
1. A general-purpose [Visual Studio credential provider](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/VisualStudioCredentialProvider.cs), part of the NuGet Client code, which opens a window allowing users to provide credentials. This is useful, for example, when accessing a private NuGet feed, such as the GitHub packages registry.
1. [Secure credential providers](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Credentials/SecurePluginCredentialProviderBuilder.cs) from the NuGet plugins directory.

In the .NET SDK and NuGet.exe, NuGet only invokes credential providers from the NuGet plugins directory.

Does that answer your question, Nikolche? If yes, I'm happy to add these details to the spec.

Reply:
My question is about plugins within the same category.

Like for example, if there are 2 plugins on the path. Which one will get invoked first?

Reply:
[Looking at the code](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Protocol/Plugins/PluginDiscoveryUtility.cs#L79-L101), I don't see any ordering. The implementation just invokes `Directory.GetDirectories` to discover plugins.